src/utils/buildConceptHierarchy.ts

```diff
@@ -0,0 +1,31 @@
+import { TFamilyConcept } from "@/types";
+
+export type TFamilyConceptTreeNode = TFamilyConcept & {
+  children: TFamilyConceptTreeNode[];
+};
+
+function buildTree(concept: TFamilyConcept, allConcepts: TFamilyConcept[], visited: Set<string>): TFamilyConceptTreeNode {
+  // Prevent infinite cycles
+  if (visited.has(concept.id)) {
+    throw new Error(`Cycle detected at concept id: ${concept.id}`);
+  }
+  visited.add(concept.id);
+
+  // find concepts where this concept is a parent and of the same type
+  const children = allConcepts
+    .filter((child) => child.subconcept_of_labels.includes(concept.preferred_label) && child.type === concept.type)
```

Comment:
Observation: Pretty good. Could be performance heavy with a lot of concepts, but I don't think we'll ever have enough for this to be significant.

Reply:
Yeh defo a good shout. Trouble is the requirement is "infinite depth" so our hands are tied to an extent. Most I've seen is around 4/5 so its not end of world.